vimiv/imutils/metadata.py

```diff
@@ -0,0 +1,418 @@
+        except piexif.InvalidImageDataError:  # File is not a jpg
+            _logger.debug("File format for '%s' does not support metadata", dest)
+        except ValueError:
+            _logger.debug("No metadata data in '%s'", dest)
+
+    def get_date_time(self) -> str:
+        with contextlib.suppress(
+            piexif.InvalidImageDataError, FileNotFoundError, KeyError
+        ):
+            return self._metadata["0th"][piexif.ImageIFD.DateTime].decode()
+        return ""
+
+
+def check_external_dependancy(handler):
+    """Decorator for ExternalKeyHandler which requires the optional pyexiv2 module.
+
+    If pyexiv2 is available, the class is left as it is. If pyexiv2 is not available
+    but the less powerful piexif module is, _ExternalKeyHandlerPiexif is returned
+    instead. If none of the two modules are available, the base implementation which
+    always throws an exception is returned.
+
+    Args:
+        handler: The class to be decorated.
+    """
+
+    if pyexiv2:
+        return handler
+
+    if piexif:
+        return _ExternalKeyHandlerPiexif
+
+    _logger.warning(
+        "There is no metadata support and therefore:\n"
+        "1. Exif/IPTC data is lost when writing images to disk.\n"
+        "2. The `:metadata` command and associated `i` keybinding is not available.\n"
+        "3. The {exif-date-time} statusbar module is not available."
+    )
+
+    return _ExternalKeyHandlerBase
+
+
+@check_external_dependancy
+class ExternalKeyHandler(_ExternalKeyHandlerBase):
+    """Main ExternalKeyHandler implementation based on pyexiv2."""
+
+    MESSAGE_SUFFIX = " by pyexiv2."
+
+    def __init__(self, filename=""):
+        super().__init__(filename)
+        try:
+            self._metadata = pyexiv2.ImageMetadata(filename)
+            self._metadata.read()
+        except FileNotFoundError:
+            _logger.debug("File %s not found", filename)
+
+    def fetch_key(self, base_key: str) -> Tuple[str, str, str]:
+        # For backwards compability, assume it has one of the following prefixes
+        for prefix in ["", "Exif.Image.", "Exif.Photo."]:
+            key = f"{prefix}{base_key}"
+            try:
+                key_name = self._metadata[key].name
+
+                try:
+                    key_value = self._metadata[key].human_value
+
+                # Not all metadata (i.e. IPTC) provide human_value, take raw_value
+                except AttributeError:
+                    value = self._metadata[key].raw_value
+
+                    # For IPTC the raw_value is a list of strings
+                    if isinstance(value, list):
+                        key_value = ", ".join(value)
+                    else:
+                        key_value = value
+
+                return (key, key_name, key_value)
+
+            except KeyError:
+                _logger.debug("Key %s is invalid for the current image", key)
+
+        return ("", "", "")
```

Comment:
Once again, re-raising the key error would save the day :blush:

Reply:
Fixed in 16ec200.